echo_server.py

```diff
@@ -0,0 +1,108 @@
+from __future__ import unicode_literals
+import socket
+import os
+from mimetypes import guess_type
+
+ROOT_DIR = "root"
+
+
+def resolve_uri(uri):
+    path = "{}{}".format(ROOT_DIR, uri)
+
+    if ".." in path:
+        raise IOError("Access Denied")
+
+    # if uri is a directory, return HTML listing of that directory as body
+    elif os.path.isdir(path):
+        directory_html = ["<li>{}</li>".format(item) for item in os.listdir(path)]
+        directory_html.insert(0, "<ul>")
+        directory_html.insert(len(directory_html), "</ul>")
+        return ("\n".join(directory_html), "text/html")
+
+    # if the resources is a file, return the contents of the file
+
+    elif os.path.isfile(path):
+        file_type = guess_type(path)[0]
+        try:
+            with open(path, 'r') as f:
+                return (f.read(), file_type)
+        except IOError:
+            raise IOError("Access Denied")
+
+    # if the requested resource cannot be found, raise an appropriate error
+    else:
+        raise IOError("File Not Found")
+
+
+def response_ok(uri):
+    response = []
+    try:
+        resolved_uri = resolve_uri(uri)
+    except IOError as excinfo:
```

Comment:
Since you want to tell the difference between different types of IO error, you might consider creating custom error classes.  It'd be better than string matching.